Improve OkHttp Docs

Andre601 opened the pull request:
Updates the complete docs in an attempt to make them more readable and organized.
This should not cause conflicts with other programs/systems (i.e. Javadoc generation, but I won't guarantee this, so testing it first is highly recommendet.

Please see #6565 for full context and actual major changes.

closes #6565

swankjesse commented:
I’d like to preserve the existing URLs, even though these URLs are slightly nicer. Keeping other peoples’ links working is nice, and there’s like a million things that link to our docs.

Andre601 commented:
Maybe re-add the old files, but the file's content informs about the move, rather than having a duplicate

swankjesse commented:
Can we keep the new structure but the old paths?

Andre601 commented:
Not possible to my knowledge

Andre601 commented:
I now made following Changes:

- Added MkDocs Redirect plugin. This plugin will create redirects for pages, allowing the old paths to now redirect to the new ones.
  All old paths have been added as a redirect to the new ones now.
  Note that in order for it to work locally will you need to install the plugin first:
  ```
  pip install mkdocs-redirects
  ```
- Fixed invalid links to other pages and images. Maybe consider to use full URLs for at least the images?

This change should now make the old links work again, which was tested. Note that I didn't re-add the placeholder files this time with the exception of the index one, so the live-preview will have a few broken pages.

swankjesse commented:
Nice find on the redirect plugin!

Andre601 commented:
I've made some new changes:

- Moved images and CSS into a dedicated `assets` directory (More clean).
- Added a `.icons/brands` directory in `assets` containing two SVG icons (kotlin and Java logo). They are from simpleicons and a license was added too.
- These SVG icons are used in the Java and Kotlin tabs to make them a bit more pretty. The Recipes section is probs best example of it.

Note that there seems to be a minor issue with the pymdownx.emoji extension where it won't render the SVG icons (`:brands-kotlin:` and `:brands-java:` when not all pages are available, so it probs requires these docs here to be generated last.

Andre601 commented:
Updated `deploy_website.sh`.

It should now copy `changelog.md` and `contributing.md` to their new respective folders.
I hope this will work as expected, so that mkdocs won't have issues with missing or wrong placed files.

yschimke commented:
https://www.andre601.ch/okhttp/

<img width="489" alt="image" src="https://user-images.githubusercontent.com/231923/110096164-8f238180-7d95-11eb-831f-13d3e5455ea1.png">

Andre601 commented:
> https://www.andre601.ch/okhttp/
> 
> <img alt="image" width="489" src="https://user-images.githubusercontent.com/231923/110096164-8f238180-7d95-11eb-831f-13d3e5455ea1.png">

Is something with that?

yschimke commented:
> As always is the live-version available to see the changes in action: https://andre601.ch/okhttp

Showing "Hi" at the live-version page.

I assumed this was a bug.

Andre601 commented:
> > As always is the live-version available to see the changes in action: https://andre601.ch/okhttp
> 
> Showing "Hi" at the live-version page.
> 
> I assumed this was a bug.

I mentioned this in the linked Issue:
> Another note:
With exception of a placeholder index file for the main page do some other pages not exist due to the nature of OkHttp's Documentation structure.

Andre601 commented:
Any update on this?
It has been open for almost a year now and would like to know if there is anything I need to change, update, etc.

Andre601 commented:
I'm considering doing some extra work on this stuff here.

- First would I go and add a theme switch, so that people could choose if they want to use the light or dark theme (which one is used by default would depend on their general system settings)
- Next would I add some extra CSS which I found from other docs that in the end would add a `external link` icon for every page that is not part of the default docs (Would essentially lead to an external page)

It is a bit sad that this PR hasn't been looked into at all for months now. I do understand that the lib itself has the main priority, but I cannot understand why there isn't a simple response here about the current status... It makes this PR feel abandoned by you.

yschimke commented:
That looks really good, I think we should go with it.  cc @swankjesse who knows more about the docs process.

I would definitely say, let's get this in before starting more work on this, so perhaps another PR that would build on this.

One nit: is there a broken link for the latest changelog? https://www.andre601.ch/okhttp/changelogs/changelog/

Andre601 commented:
> One nit: is there a broken link for the latest changelog? https://www.andre601.ch/okhttp/changelogs/changelog/

I mentioned this in the linked Issue:
> [With exception of a placeholder index file for the main page do some other pages not exist due to the nature of OkHttp's Documentation structure.](https://github.com/square/okhttp/issues/6565#issue-812922863)

The changelog file is a 404 because it doesn't seem to exist in the default docs directory. At least when I forked and updated the stuff was this the case.
I'm not sure if this is due to some of the automatic build setup you use that this is the case, but I didn't want to add a file to not break stuff.

Andre601 commented:
> I'm considering doing some extra work on this stuff here.
> 
> * First would I go and add a theme switch, so that people could choose if they want to use the light or dark theme (which one is used by default would depend on their general system settings)
> * Next would I add some extra CSS which I found from other docs that in the end would add a `external link` icon for every page that is not part of the default docs (Would essentially lead to an external page)

I now pushed the mentioned changes which should also be visible in a few minutes on the live-demo at https://andre601.ch/okhttp

swankjesse commented:
This is great. I want to apologize for not giving this any oxygen for a full year; that’s on me for spreading myself thin and should not detract from my enthusiasm for this feature.

swankjesse commented on docs/recipes.md:
````diff
@@ -559,370 +564,365 @@ It is an error to have multiple caches accessing the same cache directory simult
+      }
+    ```
+
+To prevent a response from using the cache, use [`CacheControl.FORCE_NETWORK`](http://square.github.io/okhttp/4.x/okhttp/okhttp3/-cache-control/-f-o-r-c-e_-n-e-t-w-o-r-k/). To prevent it from using the network, use [`CacheControl.FORCE_CACHE`](http://square.github.io/okhttp/4.x/okhttp/okhttp3/-cache-control/-f-o-r-c-e_-c-a-c-h-e/). Be warned: if you use `FORCE_CACHE` and the response requires the ...
````
For follow-up, let’s change these URLs back to HTTPS

swankjesse reviewed (approved):
What’s up with the `okhttp-hpacktests` submodule change?

swankjesse commented:
Revert the off-topic hpacktests change: https://github.com/square/okhttp/pull/7056

swankjesse commented:
A few notes on landing this:

 * Search broke. Material for Mkdocs only includes the search plugin if no other plugins are added. I fixed this.
 * The ‘external link’ icon isn’t working and I’d like to get rid of it. Tracking here: https://github.com/square/okhttp/issues/7060
 
 Thanks again!

Andre601 commented:
V3 docs also seem broken

yschimke commented:
Finally, the long arctic winter of Light Mode is over!
